File: src/features/bills/ui/billFetchForm.ts
export interface BillFetchForm {
  readonly values: Readonly<Record<string, string>>;
  readonly amountText: string;
}

export const emptyForm = (): BillFetchForm => ({ values: {}, amountText: '' });

export const valueOf = (form: BillFetchForm, key: string): string => form.values[key] ?? '';

export const settingValue = (form: BillFetchForm, key: string, value: string): BillFetchForm => ({
  ...form,
  values: { ...form.values, [key]: value },
});

export const settingAmount = (form: BillFetchForm, amountText: string): BillFetchForm => ({
  ...form,
  amountText,
});

export const amountPaise = (form: BillFetchForm): number | null => {
  const rupees = Number.parseFloat(form.amountText);
  if (!Number.isFinite(rupees) || rupees <= 0) return null;
  return Math.round(rupees * 100);
};

File: lib/features/bills/ui/bill_fetch_form.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'bill_fetch_form.freezed.dart';

@freezed
abstract class BillFetchForm with _$BillFetchForm {
  const BillFetchForm._();

  const factory BillFetchForm({
    required Map<String, String> values,
    required String amountText,
  }) = _BillFetchForm;

  factory BillFetchForm.empty() =>
      const BillFetchForm(values: {}, amountText: '');

  String valueOf(String key) => values[key] ?? '';

  BillFetchForm setting(String key, String value) =>
      copyWith(values: {...values, key: value});

  int? get amountPaise {
    final rupees = double.tryParse(amountText);
    if (rupees == null || rupees <= 0) return null;
    return (rupees * 100).round();
  }
}
